cleanlab/internal/neighbor/knn_graph.py

```diff
@@ -106,6 +120,12 @@ def construct_knn_graph_from_index(knn: NearestNeighbors) -> csr_matrix:
 
     distances, indices = knn.kneighbors(return_distance=True)
 
+    if correct_exact_duplicates:
+        # NOTE: Accessing the features used to fit the knn object is a short-term solution.
+        # WARNING: Accessing an object's private attribute from a third-party library is
+        _features = knn._fit_X
```

Comment:
why do you want to have such a dangerous line of code inside of `construct_knn_graph_from_index()`?  

Where are you using `construct_knn_graph_from_index()` in a place that `features` is not also directly accessible?

Eg. here `features` still seems accessible:

https://github.com/cleanlab/cleanlab/blob/b063279788b636484709302a3b1e79c4bfe30597/cleanlab/internal/neighbor/knn_graph.py#L191-L193

I see you use this method in `issue_manager/outlier`:  https://github.com/cleanlab/cleanlab/blob/b063279788b636484709302a3b1e79c4bfe30597/cleanlab/datalab/internal/issue_manager/outlier.py#L230
where `features` is not accessible. But that doesn't mean we should invoke the dangerous `._fit_X`  everywhere, we should instead only invoke it where absolutely necessary (ie. in outliers only).

My suggestion: make `construct_knn_graph_from_index()` (or a parent method that calls it) able to take in optional arg `features`, and rely on `features` if available instead of `._fit_X`.  Make sure `._fix_X` is only relied on in outlier-code, and explicitly confirm it works across all recent sklearn versions that Cleanlab package supports.

Reply:
Updated `construct_knn_graph_from_index()` to take in `correction_features` as an optional argument.
It's a bad idea overall to mix the variable names `knn` and `features` when the `features` are not used by `knn` to do the search.

Reply:
thanks!  Please confirm the following statement:  

Now Outliers would be the only issue manager that would be skipped if sklearn updates to no longer allow `._fit_X` attribute (and in that case the exception would be caught and Datalab would just skip outlier detection).   All of the other issue managers would still work.

Reply:
Thanks for the question!

**Short Answer**: 🎉 _**Yes**_ 🎉 , the outlier issue manager would be the only one affected by an update to sklearn that removes the ._fit_X attribute, causing it to be skipped if the exception is caught.

**Detailed Response**:

To provide a more complete picture, there is only one place in our codebase where we use the `._fit_X` attribute, which is in the `OutOfDistribution` class. Based on this, we can confirm the following:

- The outlier issue manager is indeed the only manager that calls `OutOfDistribution` for scoring features.
- This process necessitates the construction of a k-NN graph (or an equivalent data structure) and subsequent correction using the `._fit_X` attribute, which we have tried to avoid.

It's important to note that the outlier issue manager can avoid being skipped if just gets a k-NN graph. In this case, it would not call `OutOfDistribution`, thereby bypassing the need for the `._fit_X` attribute.

Also, our test suite will catch if sklearn removes the ._fit_X attribute. This can be verified in our test file [here](https://github.com/cleanlab/cleanlab/blob/c13f32a2a6c1d6f3166d760b60f805c94d9f54fe/tests/internal/neighbor/test_neighbor.py#L52).

---

In summary, while the statement is correct under the condition that the outlier issue manager calls `OutOfDistribution`, there is a potential workaround by directly providing a k-NN graph to the outlier issue manager.

I hope this clarifies the situation.